Sources/SourceKitLSP/IndexProgressManager.swift

```diff
@@ -70,23 +64,47 @@ actor IndexProgressManager {
       workDoneProgress = nil
       return
     }
-    var scheduled: [DocumentURI] = []
-    var executing: [DocumentURI] = []
+    var isGeneratingBuildGraph = false
+    var indexTasks: [DocumentURI: IndexTaskStatus] = [:]
+    var preparationTasks: [ConfiguredTarget: IndexTaskStatus] = [:]
     for indexManager in await sourceKitLSPServer.workspaces.compactMap({ $0.semanticIndexManager }) {
-      let inProgress = await indexManager.inProgressIndexFiles
-      scheduled += inProgress.scheduled
-      executing += inProgress.executing
+      let inProgress = await indexManager.inProgressTasks
+      isGeneratingBuildGraph = isGeneratingBuildGraph || inProgress.isGeneratingBuildGraph
+      indexTasks.merge(inProgress.indexTasks) { lhs, rhs in
+        return max(lhs, rhs)
+      }
+      preparationTasks.merge(inProgress.preparationTasks) { lhs, rhs in
+        return max(lhs, rhs)
+      }
     }
 
-    if scheduled.isEmpty && executing.isEmpty {
+    if indexTasks.isEmpty {
       // Nothing left to index. Reset the target count and dismiss the work done progress.
       queuedIndexTasks = 0
       workDoneProgress = nil
       return
     }
 
-    let finishedTasks = queuedIndexTasks - scheduled.count - executing.count
-    let message = "\(finishedTasks) / \(queuedIndexTasks)"
+    // We can get into a situation where queuedIndexTasks < indexTasks.count if we haven't processed all
+    // `indexTasksWereScheduled` calls yet but the semantic index managers already track them in their in-progress tasks.
+    // Clip the finished tasks to 0 because showing a negative number there looks stupid.
+    let finishedTasks = max(queuedIndexTasks - indexTasks.count, 0)
```

Comment:
Just realized that `queuedIndexTasks` will be artificially high if you schedule a re-index for a file multiple times, right?